File: outils/outil-B-preuve-image/preuve_image/quantification.py
"""
quantification.py — Les tables de quantification d'un JPEG (§16, §20).

CE QUE CE MODULE ÉTABLIT, ET CE QU'IL N'ÉTABLIT PAS
───────────────────────────────────────────────────
Une image JPEG porte, dans ses segments DQT, les tables qui ont servi à la
quantifier. Ce sont des données EXACTES, présentes dans tout JPEG, et elles
survivent à la purge de l'EXIF : un fichier dont toutes les métadonnées ont été
retirées porte encore ses tables. C'est ce qui en fait un objet intéressant.

CE QUI EST ÉTABLI :
  · les tables elles-mêmes, telles qu'écrites, et leur empreinte ;
  · le facteur de qualité au sens de la bibliothèque de référence (IJG), quand
    les tables en dérivent — avec l'écart résiduel, qui dit à quel point ;
  · le nombre de tables, leur précision, et le sous-échantillonnage chromatique
    lu dans le segment SOF ;
  · le fait qu'une image a été RÉENCODÉE quand ses tables ne correspondent à
    aucun facteur IJG entier, ce qui est le cas des encodeurs propriétaires.

CE QUI N'EST PAS ÉTABLI, ET NE PEUT PAS L'ÊTRE ICI :
  · la MARQUE de l'appareil ou du logiciel. Associer une signature de
    quantification à « Canon DIGIC » ou « algorithme WhatsApp » demande un
    corpus de fichiers réels, collectés et vérifiés, appareil par appareil et
    version par version. Ce paquet n'en a pas. Un registre vide et extensible
    est fourni — `SIGNATURES_CONNUES` — plutôt qu'une table inventée : une
    correspondance affirmée sans corpus derrière serait une conjecture
    présentée comme un fait, et c'est exactement ce que le protocole interdit.

L'empreinte des tables est donc rendue pour être COMPARÉE : deux fichiers dont
les tables ont la même empreinte sont sortis de la même chaîne d'encodage, aux
mêmes réglages. C'est un rapprochement entre deux fichiers qu'on a tous les
deux, jamais une identification contre une base qu'on n'a pas.
"""

import hashlib
import struct
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

__all__ = [
    "QuantificationError",
    "TableQuantification",
    "AnalyseQuantification",
    "analyser_quantification",
    "TABLE_LUMINANCE_ANNEXE_K",
    "TABLE_CHROMINANCE_ANNEXE_K",
    "table_ijg",
    "qualite_ijg_estimee",
    "SIGNATURES_CONNUES",
    "MOTIF_AUCUNE_SIGNATURE",
]


class QuantificationError(ValueError):
    """Flux JPEG malformé, ou dépourvu de table de quantification."""


#: Table de luminance de l'annexe K de la norme ISO/IEC 10918-1, en ordre
#: zigzag. C'est la table dont dérivent, par mise à l'échelle, celles de la
#: quasi-totalité des encodeurs — la référence de l'IJG (libjpeg) comme celle
#: de la plupart des appareils.
TABLE_LUMINANCE_ANNEXE_K = (
    16, 11, 12, 14, 12, 10, 16, 14, 13, 14, 18, 17, 16, 19, 24, 40,
    26, 24, 22, 22, 24, 49, 35, 37, 29, 40, 58, 51, 61, 60, 57, 51,
    56, 55, 64, 72, 92, 78, 64, 68, 87, 69, 55, 56, 80, 109, 81, 87,
    95, 98, 103, 104, 103, 62, 77, 113, 121, 112, 100, 120, 92, 101, 103, 99,
)

#: Table de chrominance de la même annexe.
TABLE_CHROMINANCE_ANNEXE_K = (
    17, 18, 18, 24, 21, 24, 47, 26, 26, 47, 99, 66, 56, 66, 99, 99,
    99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
    99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
    99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
)

#: Le registre des signatures reconnues. VIDE, et c'est délibéré.
#:
#: Le remplir demande un corpus de fichiers réels, collectés appareil par
#: appareil et version par version, dont on puisse dire d'où ils viennent.
#: Chaque entrée serait : empreinte des tables → (producteur, ce qui l'établit).
#: Y écrire des correspondances de mémoire produirait des identifications
#: fausses présentées comme des faits — et une identification fausse dans un
#: dossier probatoire coûte plus cher que pas d'identification du tout.
SIGNATURES_CONNUES: Dict[str, Tuple[str, str]] = {}

MOTIF_AUCUNE_SIGNATURE = (
    "Aucun rapprochement : le registre des signatures est vide. Le remplir "
    "demande un corpus de fichiers réels dont la provenance est établie, "
    "appareil par appareil et version par version. L'empreinte ci-dessus reste "
    "utilisable pour COMPARER deux fichiers que vous avez tous les deux — "
    "des tables identiques sortent de la même chaîne d'encodage aux mêmes "
    "réglages — mais elle n'identifie rien contre une base qu'on n'a pas."
)


@dataclass(frozen=True)
class TableQuantification:
    """Une table DQT, telle qu'elle est écrite dans le fichier.

    `valeurs` est en ordre ZIGZAG, celui du fichier, et non en ordre de
    balayage : les réordonner ici ferait diverger l'empreinte de celle qu'un
    autre outil calculerait sur les mêmes octets.
    """

    identifiant: int
    precision_bits: int
    valeurs: Tuple[int, ...]
    offset: int

    @property
    def empreinte(self) -> str:
        return hashlib.sha256(
            ",".join(str(v) for v in self.valeurs).encode("ascii")
        ).hexdigest()

    @property
    def somme(self) -> int:
        """La somme des coefficients. Grossière, mais elle ordonne les qualités."""
        return sum(self.valeurs)


@dataclass(frozen=True)
class AnalyseQuantification:
    """Ce que les tables d'un JPEG permettent de dire."""

    tables: Tuple[TableQuantification, ...]
    #: L'empreinte de l'ENSEMBLE des tables, dans l'ordre du fichier. C'est
    #: elle qui sert à rapprocher deux fichiers.
    empreinte_ensemble: str
    #: Le facteur de qualité IJG le plus proche, et l'écart qui reste.
    qualite_ijg: Optional[int]
    #: Nombre de coefficients qui diffèrent de la table IJG reconstruite.
    ecart_a_ijg: Optional[int]
    #: Vrai quand les tables coïncident EXACTEMENT avec un facteur IJG entier.
    conforme_ijg: bool
    #: Le sous-échantillonnage chromatique lu dans le SOF, ex. « 4:2:0 ».
    sous_echantillonnage: Optional[str]
    #: Progressif ou séquentiel : un réencodage change souvent ce choix.
    progressif: Optional[bool]
    largeur: Optional[int] = None
    hauteur: Optional[int] = None
    composantes: int = 0
    #: Le rapprochement, s'il y en a un. Voir SIGNATURES_CONNUES.
    signature: Optional[Tuple[str, str]] = None
    motif_absence_signature: str = MOTIF_AUCUNE_SIGNATURE
    marqueurs: Tuple[str, ...] = field(default_factory=tuple)


def table_ijg(qualite: int, base: Tuple[int, ...] = TABLE_LUMINANCE_ANNEXE_K) -> Tuple[int, ...]:
    """La table que produit la bibliothèque de référence IJG pour ce facteur.

    L'algorithme est celui de `jpeg_set_quality` : le facteur est converti en
    une échelle, puis chaque coefficient de la table de l'annexe K est mis à
    l'échelle et borné à [1 ; 255]. Il est reproduit ici plutôt qu'appelé, pour
    que la comparaison ne dépende d'aucune bibliothèque installée.
    """
    if not 1 <= qualite <= 100:
        raise QuantificationError("Le facteur de qualité IJG est compris entre 1 et 100.")
    echelle = 5000 // qualite if qualite < 50 else 200 - qualite * 2
    return tuple(
        min(255, max(1, (v * echelle + 50) // 100))
        for v in base
    )


def qualite_ijg_estimee(
    valeurs: Tuple[int, ...], base: Tuple[int, ...] = TABLE_LUMINANCE_ANNEXE_K
) -> Tuple[Optional[int], Optional[int]]:
    """Le facteur IJG dont la table est la plus proche, et l'écart qui reste.

    L'écart est le nombre de coefficients qui DIFFÈRENT, pas une distance :
    zéro veut dire que la table est exactement celle de ce facteur, et tout le
    reste veut dire qu'elle n'en vient pas — ce qui est le cas des encodeurs
    d'appareils photo et de plusieurs services de messagerie.

    Rendre un facteur « approché » sans cet écart laisserait croire à une
    identification là où il n'y a qu'une ressemblance.
    """
    if len(valeurs) != 64:
        return None, None
    meilleur, ecart_min = None, None
    for q in range(1, 101):
        candidate = table_ijg(q, base)
        ecart = sum(1 for a, b in zip(valeurs, candidate) if a != b)
        if ecart_min is None or ecart < ecart_min:
            meilleur, ecart_min = q, ecart
            if ecart == 0:
                break
    return meilleur, ecart_min


#: Les marqueurs JPEG qu'on nomme. Leur présence renseigne sur la chaîne : un
#: JPEG progressif sort rarement d'un appareil photo.
_MARQUEURS = {
    0xC0: "SOF0 (séquentiel, Huffman)",
    0xC1: "SOF1 (séquentiel étendu)",
    0xC2: "SOF2 (progressif)",
    0xC3: "SOF3 (sans perte)",
    0xC4: "DHT (tables de Huffman)",
    0xC9: "SOF9 (arithmétique)",
    0xCC: "DAC (codage arithmétique)",
    0xDB: "DQT (tables de quantification)",
    0xDD: "DRI (intervalle de redémarrage)",
    0xDA: "SOS (début du balayage)",
    0xE0: "APP0 (JFIF)",
    0xE1: "APP1 (EXIF ou XMP)",
    0xE2: "APP2 (ICC ou MPF)",
    0xEC: "APP12 (Ducky, Picture Info)",
    0xED: "APP13 (Photoshop IRB)",
    0xEE: "APP14 (Adobe)",
    0xFE: "COM (commentaire)",
}


def _sous_echantillonnage(composantes: List[Tuple[int, int, int]]) -> Optional[str]:
    """Déduit la notation « 4:x:x » des facteurs d'échantillonnage du SOF.

    La notation usuelle se lit sur les facteurs de la composante de luminance,
    rapportés à ceux de la chrominance. Elle n'a de sens qu'à trois composantes ;
    au-delà ou en deçà, on rend None plutôt qu'une notation qui n'existe pas.
    """
    if len(composantes) != 3:
        return None
    hy, vy = composantes[0][1], composantes[0][2]
    hc, vc = composantes[1][1], composantes[1][2]
    if hc == 0 or vc == 0:
        return None
    rh, rv = hy // hc, vy // vc
    return {
        (1, 1): "4:4:4", (2, 1): "4:2:2", (2, 2): "4:2:0",
        (1, 2): "4:4:0", (4, 1): "4:1:1", (4, 2): "4:1:0",
    }.get((rh, rv))


def analyser_quantification(donnees: bytes) -> AnalyseQuantification:
    """Lit les tables DQT d'un JPEG et ce que le SOF déclare.

    Le balayage s'arrête au SOS : au-delà commencent les données comprimées,
    où toute séquence d'octets peut ressembler à un marqueur.
    """
    if len(donnees) < 4 or donnees[0] != 0xFF or donnees[1] != 0xD8:
        raise QuantificationError("Fichier non reconnu comme JPEG (SOI absent).")

    tables: List[TableQuantification] = []
    marqueurs: List[str] = []
    composantes: List[Tuple[int, int, int]] = []
    largeur = hauteur = None
    progressif = None

    pos = 2
    while pos + 4 <= len(donnees):
        if donnees[pos] != 0xFF:
            break
        m = donnees[pos + 1]
        if m == 0xD8 or 0xD0 <= m <= 0xD7:
            pos += 2
            continue
        if m == 0xD9:
            break
        longueur = struct.unpack_from(">H", donnees, pos + 2)[0]
        if longueur < 2 or pos + 2 + longueur > len(donnees):
            break
        corps = donnees[pos + 4 : pos + 2 + longueur]
        nom = _MARQUEURS.get(m, f"0x{m:02X}")
        if nom not in marqueurs:
            marqueurs.append(nom)

        if m == 0xDB:
            # Un même segment DQT peut porter plusieurs tables à la suite.
            p = 0
            while p < len(corps):
                entete = corps[p]
                precision = 16 if (entete >> 4) else 8
                identifiant = entete & 0x0F
                n = 64 * (2 if precision == 16 else 1)
                if p + 1 + n > len(corps):
                    break
                brut = corps[p + 1 : p + 1 + n]
                if precision == 16:
                    valeurs = struct.unpack(f">{64}H", brut)
                else:
                    valeurs = tuple(brut)
                tables.append(TableQuantification(
                    identifiant=identifiant, precision_bits=precision,
                    valeurs=tuple(valeurs), offset=pos + 4 + p,
                ))
                p += 1 + n
        elif m in (0xC0, 0xC1, 0xC2, 0xC3, 0xC9) and len(corps) >= 6:
            progressif = m == 0xC2
            hauteur, largeur = struct.unpack_from(">HH", corps, 1)
            nb = corps[5]
            for i in range(nb):
                q = 6 + i * 3
                if q + 3 > len(corps):
                    break
                composantes.append((corps[q], corps[q + 1] >> 4, corps[q + 1] & 0x0F))
        elif m == 0xDA:
            break

        pos += 2 + longueur

    if not tables:
        raise QuantificationError(
            "Aucune table de quantification (DQT) trouvée avant le début du balayage. "
            "Un JPEG en porte toujours : le fichier est tronqué, ou ce n'en est pas un."
        )

    # L'empreinte porte sur TOUTES les tables, dans l'ordre du fichier : deux
    # images de même table de luminance mais de chrominance différente ne
    # viennent pas de la même chaîne, et les confondre serait un faux
    # rapprochement.
    ensemble = ";".join(
        f"{t.identifiant}:{t.precision_bits}:" + ",".join(str(v) for v in t.valeurs)
        for t in tables
    )
    empreinte_ensemble = hashlib.sha256(ensemble.encode("ascii")).hexdigest()

    # Le facteur IJG se lit sur la table de luminance, celle d'identifiant 0.
    luminance = next((t for t in tables if t.identifiant == 0), tables[0])
    qualite, ecart = qualite_ijg_estimee(luminance.valeurs)

    return AnalyseQuantification(
        tables=tuple(tables),
        empreinte_ensemble=empreinte_ensemble,
        qualite_ijg=qualite,
        ecart_a_ijg=ecart,
        conforme_ijg=ecart == 0,
        sous_echantillonnage=_sous_echantillonnage(composantes),
        progressif=progressif,
        largeur=largeur,
        hauteur=hauteur,
        composantes=len(composantes),
        signature=SIGNATURES_CONNUES.get(empreinte_ensemble),
        marqueurs=tuple(marqueurs),
    )

File: src/lib/preuve-image/quantification.ts
/**
 * quantification.ts — Port TypeScript de `preuve_image.quantification` (outil B).
 *
 * CE FICHIER N'EST PAS LA RÉFÉRENCE.
 * La référence est `outils/outil-B-preuve-image/preuve_image/quantification.py`.
 * Ce port est épinglé au Python par `vecteurs-or-extraction.json`, que
 * `scripts/verifier-port-extraction.mjs` rejoue ici.
 *
 * CE QUE CE MODULE ÉTABLIT, ET CE QU'IL N'ÉTABLIT PAS
 * ───────────────────────────────────────────────────
 * Une image JPEG porte, dans ses segments DQT, les tables qui ont servi à la
 * quantifier. Ce sont des données exactes, présentes dans tout JPEG, et elles
 * SURVIVENT à la purge de l'EXIF : un fichier dont toutes les métadonnées ont
 * été retirées porte encore ses tables.
 *
 * Établi : les tables, leur empreinte, le facteur de qualité IJG quand elles en
 * dérivent — avec l'écart résiduel —, le sous-échantillonnage et le mode.
 *
 * NON établi : la marque de l'appareil ou du logiciel. Associer une signature à
 * « Canon DIGIC » ou « algorithme WhatsApp » demande un corpus de fichiers
 * réels dont la provenance est établie. Ce paquet n'en a pas, et le registre
 * `SIGNATURES_CONNUES` est vide plutôt qu'inventé.
 */

export class QuantificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'QuantificationError';
  }
}

/**
 * Table de luminance de l'annexe K d'ISO/IEC 10918-1, en ordre zigzag. C'est
 * la table dont dérivent, par mise à l'échelle, celles de la quasi-totalité
 * des encodeurs.
 */
export const TABLE_LUMINANCE_ANNEXE_K: readonly number[] = [
  16, 11, 12, 14, 12, 10, 16, 14, 13, 14, 18, 17, 16, 19, 24, 40,
  26, 24, 22, 22, 24, 49, 35, 37, 29, 40, 58, 51, 61, 60, 57, 51,
  56, 55, 64, 72, 92, 78, 64, 68, 87, 69, 55, 56, 80, 109, 81, 87,
  95, 98, 103, 104, 103, 62, 77, 113, 121, 112, 100, 120, 92, 101, 103, 99,
];

export const TABLE_CHROMINANCE_ANNEXE_K: readonly number[] = [
  17, 18, 18, 24, 21, 24, 47, 26, 26, 47, 99, 66, 56, 66, 99, 99,
  99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
  99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
  99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99,
];

/**
 * Le registre des signatures reconnues. VIDE, et c'est délibéré.
 *
 * Le remplir demande un corpus de fichiers réels, collectés appareil par
 * appareil et version par version. Y écrire des correspondances de mémoire
 * produirait des identifications fausses présentées comme des faits — et une
 * identification fausse dans un dossier probatoire coûte plus cher que pas
 * d'identification du tout.
 */
export const SIGNATURES_CONNUES: Record<string, [string, string]> = {};

export const MOTIF_AUCUNE_SIGNATURE =
  'Aucun rapprochement : le registre des signatures est vide. Le remplir '
  + "demande un corpus de fichiers réels dont la provenance est établie, "
  + 'appareil par appareil et version par version. L’empreinte ci-dessus reste '
  + 'utilisable pour COMPARER deux fichiers que vous avez tous les deux — '
  + 'des tables identiques sortent de la même chaîne d’encodage aux mêmes '
  + 'réglages — mais elle n’identifie rien contre une base qu’on n’a pas.';

export interface TableQuantification {
  identifiant: number;
  precisionBits: number;
  /** En ordre ZIGZAG, celui du fichier : les réordonner ferait diverger l'empreinte. */
  valeurs: number[];
  offset: number;
}

export interface AnalyseQuantification {
  tables: TableQuantification[];
  /** L'empreinte de l'ENSEMBLE des tables, dans l'ordre du fichier. */
  empreinteEnsemble: string;
  qualiteIjg: number | null;
  /** Nombre de coefficients qui diffèrent de la table IJG reconstruite. */
  ecartAIjg: number | null;
  conformeIjg: boolean;
  sousEchantillonnage: string | null;
  progressif: boolean | null;
  largeur: number | null;
  hauteur: number | null;
  composantes: number;
  signature: [string, string] | null;
  motifAbsenceSignature: string;
  marqueurs: string[];
}

/**
 * La table que produit la bibliothèque de référence IJG pour ce facteur.
 *
 * L'algorithme est celui de `jpeg_set_quality`, reproduit ici plutôt
 * qu'emprunté, pour que la comparaison ne dépende d'aucune bibliothèque.
 */
export function tableIjg(
  qualite: number, base: readonly number[] = TABLE_LUMINANCE_ANNEXE_K,
): number[] {
  if (!(qualite >= 1 && qualite <= 100)) {
    throw new QuantificationError('Le facteur de qualité IJG est compris entre 1 et 100.');
  }
  // Division ENTIÈRE des deux côtés : `Math.floor` là où Python écrit `//`.
  const echelle = qualite < 50 ? Math.floor(5000 / qualite) : 200 - qualite * 2;
  return base.map((v) => Math.min(255, Math.max(1, Math.floor((v * echelle + 50) / 100))));
}

/**
 * Le facteur IJG dont la table est la plus proche, et l'écart qui reste.
 *
 * L'écart est le nombre de coefficients qui DIFFÈRENT, pas une distance : zéro
 * veut dire que la table est exactement celle de ce facteur, et tout le reste
 * veut dire qu'elle n'en vient pas. Rendre un facteur « approché » sans cet
 * écart laisserait croire à une identification là où il n'y a qu'une
 * ressemblance.
 */
export function qualiteIjgEstimee(
  valeurs: readonly number[], base: readonly number[] = TABLE_LUMINANCE_ANNEXE_K,
): [number | null, number | null] {
  if (valeurs.length !== 64) return [null, null];
  let meilleur: number | null = null;
  let ecartMin: number | null = null;
  for (let q = 1; q <= 100; q++) {
    const candidate = tableIjg(q, base);
    let ecart = 0;
    for (let i = 0; i < 64; i++) if (valeurs[i] !== candidate[i]) ecart += 1;
    if (ecartMin === null || ecart < ecartMin) {
      meilleur = q;
      ecartMin = ecart;
      if (ecart === 0) break;
    }
  }
  return [meilleur, ecartMin];
}

/** Les marqueurs JPEG qu'on nomme. Un JPEG progressif sort rarement d'un appareil. */
const MARQUEURS: Record<number, string> = {
  0xc0: 'SOF0 (séquentiel, Huffman)',
  0xc1: 'SOF1 (séquentiel étendu)',
  0xc2: 'SOF2 (progressif)',
  0xc3: 'SOF3 (sans perte)',
  0xc4: 'DHT (tables de Huffman)',
  0xc9: 'SOF9 (arithmétique)',
  0xcc: 'DAC (codage arithmétique)',
  0xdb: 'DQT (tables de quantification)',
  0xdd: 'DRI (intervalle de redémarrage)',
  0xda: 'SOS (début du balayage)',
  0xe0: 'APP0 (JFIF)',
  0xe1: 'APP1 (EXIF ou XMP)',
  0xe2: 'APP2 (ICC ou MPF)',
  0xec: 'APP12 (Ducky, Picture Info)',
  0xed: 'APP13 (Photoshop IRB)',
  0xee: 'APP14 (Adobe)',
  0xfe: 'COM (commentaire)',
};

const NOTATIONS: Record<string, string> = {
  '1,1': '4:4:4', '2,1': '4:2:2', '2,2': '4:2:0',
  '1,2': '4:4:0', '4,1': '4:1:1', '4,2': '4:1:0',
};

/**
 * Déduit la notation « 4:x:x » des facteurs d'échantillonnage du SOF.
 * Elle n'a de sens qu'à trois composantes ; au-delà ou en deçà, on rend null
 * plutôt qu'une notation qui n'existe pas.
 */
function sousEchantillonnage(composantes: [number, number, number][]): string | null {
  if (composantes.length !== 3) return null;
  const [, hy, vy] = composantes[0];
  const [, hc, vc] = composantes[1];
  if (hc === 0 || vc === 0) return null;
  return NOTATIONS[`${Math.floor(hy / hc)},${Math.floor(vy / vc)}`] ?? null;
}

async function sha256Texte(texte: string): Promise<string> {
  const octets = new TextEncoder().encode(texte);
  const c = await crypto.subtle.digest('SHA-256', octets.slice().buffer);
  return Array.from(new Uint8Array(c)).map((b) => b.toString(16).padStart(2, '0')).join('');
}

/** L'empreinte d'une table seule, sur ses valeurs en ordre zigzag. */
export function empreinteTable(t: TableQuantification): Promise<string> {
  return sha256Texte(t.valeurs.join(','));
}

/**
 * Lit les tables DQT d'un JPEG et ce que le SOF déclare.
 *
 * Le balayage s'arrête au SOS : au-delà commencent les données comprimées, où
 * toute séquence d'octets peut ressembler à un marqueur.
 */
export async function analyserQuantification(donnees: Uint8Array): Promise<AnalyseQuantification> {
  if (donnees.length < 4 || donnees[0] !== 0xff || donnees[1] !== 0xd8) {
    throw new QuantificationError('Fichier non reconnu comme JPEG (SOI absent).');
  }
  const vue = new DataView(donnees.buffer, donnees.byteOffset, donnees.byteLength);
  const tables: TableQuantification[] = [];
  const marqueurs: string[] = [];
  const composantes: [number, number, number][] = [];
  let largeur: number | null = null;
  let hauteur: number | null = null;
  let progressif: boolean | null = null;

  let pos = 2;
  while (pos + 4 <= donnees.length) {
    if (donnees[pos] !== 0xff) break;
    const m = donnees[pos + 1];
    if (m === 0xd8 || (m >= 0xd0 && m <= 0xd7)) { pos += 2; continue; }
    if (m === 0xd9) break;
    const longueur = vue.getUint16(pos + 2, false);
    if (longueur < 2 || pos + 2 + longueur > donnees.length) break;
    const debutCorps = pos + 4;
    const finCorps = pos + 2 + longueur;
    const nom = MARQUEURS[m] ?? `0x${m.toString(16).padStart(2, '0').toUpperCase()}`;
    if (!marqueurs.includes(nom)) marqueurs.push(nom);

    if (m === 0xdb) {
      // Un même segment DQT peut porter plusieurs tables à la suite.
      let p = debutCorps;
      while (p < finCorps) {
        const entete = donnees[p];
        const precision = (entete >> 4) ? 16 : 8;
        const identifiant = entete & 0x0f;
        const n = 64 * (precision === 16 ? 2 : 1);
        if (p + 1 + n > finCorps) break;
        const valeurs: number[] = [];
        for (let i = 0; i < 64; i++) {
          valeurs.push(precision === 16
            ? vue.getUint16(p + 1 + i * 2, false)
            : donnees[p + 1 + i]);
        }
        tables.push({ identifiant, precisionBits: precision, valeurs, offset: p });
        p += 1 + n;
      }
    } else if ([0xc0, 0xc1, 0xc2, 0xc3, 0xc9].includes(m) && longueur >= 8) {
      progressif = m === 0xc2;
      hauteur = vue.getUint16(debutCorps + 1, false);
      largeur = vue.getUint16(debutCorps + 3, false);
      const nb = donnees[debutCorps + 5];
      for (let i = 0; i < nb; i++) {
        const q = debutCorps + 6 + i * 3;
        if (q + 3 > finCorps) break;
        composantes.push([donnees[q], donnees[q + 1] >> 4, donnees[q + 1] & 0x0f]);
      }
    } else if (m === 0xda) {
      break;
    }
    pos += 2 + longueur;
  }

  if (tables.length === 0) {
    throw new QuantificationError(
      'Aucune table de quantification (DQT) trouvée avant le début du balayage. '
      + "Un JPEG en porte toujours : le fichier est tronqué, ou ce n'en est pas un.",
    );
  }

  // L'empreinte porte sur TOUTES les tables, dans l'ordre du fichier : deux
  // images de même luminance mais de chrominance différente ne viennent pas de
  // la même chaîne, et les confondre serait un faux rapprochement.
  const ensemble = tables
    .map((t) => `${t.identifiant}:${t.precisionBits}:${t.valeurs.join(',')}`)
    .join(';');
  const empreinteEnsemble = await sha256Texte(ensemble);

  const luminance = tables.find((t) => t.identifiant === 0) ?? tables[0];
  const [qualite, ecart] = qualiteIjgEstimee(luminance.valeurs);

  return {
    tables,
    empreinteEnsemble,
    qualiteIjg: qualite,
    ecartAIjg: ecart,
    conformeIjg: ecart === 0,
    sousEchantillonnage: sousEchantillonnage(composantes),
    progressif,
    largeur,
    hauteur,
    composantes: composantes.length,
    signature: SIGNATURES_CONNUES[empreinteEnsemble] ?? null,
    motifAbsenceSignature: MOTIF_AUCUNE_SIGNATURE,
    marqueurs,
  };
}
